### PR TITLE
fix(ThemeProvider): update theme with prop change

### PR DESCRIPTION
**Pull request by arpitBhalla**
Related #3665

### CODE REVIEW

**Comment by Nantris** [comment]
Thanks again for all your efforts @arpitBhalla. I think that the `ThemeProvider` changes seem good to go in my view.

If merged, it should resolve: #3655 and #3665

**Comment by Nantris** [comment]
@arpitBhalla I just wanted to thank you again for your great work. Thanks so much for what you do!

**Comment by Nantris** [comment]
@arpitBhalla is it planned to merge this? It's seemed ready to go for many months but it still isn't merged, not to mention released. Could you provide some clarity?

File: example/App.tsx [hunk 1]
```diff
@@ -5,12 +5,15 @@ import { cacheImages, cacheFonts } from './src/helpers/AssetsCaching';
...
 import { SafeAreaProvider } from 'react-native-safe-area-context';
 import * as SplashScreen from 'expo-splash-screen';
+import { useColorScheme } from 'react-native';
 
 SplashScreen.preventAutoHideAsync();
 
 export default () => {
   const [isReady, setIsReady] = useState(false);
 
+  theme.mode = useColorScheme();
```
Comment (khushal87) [inline example/App.tsx]: Don't store the result of `useColorScheme()` directly to theme.mode. Rather do this:

```suggestion
  const colorScheme = useColorScheme();
  theme.mode = colorScheme;
```

This looks better, imo.

**Comment by Nantris** [comment]
Friendly bump.

I think the code changes look good. I noticed there's been no new release candidate for a while. Is it planned to release one soon?

Thank you again for all of your hard work, and please don't mistake my bump for impatience.

@arpitBhalla

**Comment by markrickert** [comment]
This PR broke my `updateTheme()` functionality when upgrading from `rc.7` to `rc.8`.

I'm using a hook to determine if fonts should be scaled in the app:
```jsx
// These constants determine mow much bigger the font size should get based on the user's
// accessibility settings. Even if they turn the dial all the way to 11, we will only ever
// scale the fonts by these factors. This is to prevent the font size from getting too large
// and completely breaking the layout.
export const MAX_FONT_SCALE = 1.2;
export const MIN_FONT_SCALE = 0.8;

// Returns fontScaling props for Text and TextInput components
export const useFontScaling = (): Partial<TextProps> => {
  const { preferenceStore } = useStores();

  const fontScaling: Partial<TextProps> = React.useMemo(() => {
    const afs = preferenceStore?.allowFontScaling || false;
    return {
      minimumFontScale: afs ? MIN_FONT_SCALE : 1, // This prevents the font from getting too small.
      maxFontSizeMultiplier: afs ? MAX_FONT_SCALE : 1, // This prevents the font from getting too big.
      allowFontScaling: afs, // This allows the font to be scaled.
    };
  }, [preferenceStore?.allowFontScaling]);

  return fontScaling;
};
```

Then whenever it changes, i try and update the current theme:

```jsx
const { updateTheme } = useTheme();
const fontScalingProps = useFontScaling();

React.useEffect(()=>{
  updateTheme({
    components: {
      Text: {
        ...fontScalingProps,
      },
      // Other rneui components as necessary
    }
  });
},[fontScalingProps]);
```

`rc.8` completely broke this functionality because of this PR. I manually reverted changes in the `ThemeProvider.tsx` and the updates resumed working as intended.

Am I updating the theme incorrectly or is this an actual regression? Some guidance would be appreciated, thanks!

**Comment by Nantris** [comment]
@markrickert it happens because you need to maintain object equality for the theme object after this PR (which if I'm not mistaken is how version 3.x handled it)

An approach to handling this might be like:

```
const [theme, setTheme] = React.useState({
    components: {
      Text: {
        ...fontScalingProps,
      },
      // Other rneui components as necessary
    }
});

React.useEffect(() => {
    const updatedTheme = {
        components: {
            Text: {
                ...fontScalingProps,
            },
            // Other rneui components as necessary
        }
    };
    
    setTheme(updatedTheme);
    updateTheme(updatedTheme);
}, [fontScalingProps]);
```

**Pull request merged**